File: 백준/Silver/12789. 도키도키 간식드리미/도키도키 간식드리미.java
import java.util.Scanner;
import java.util.Stack;

public class Main {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int n = sc.nextInt();

        int[] arr = new int[n];
        for (int i = 0; i < n; i++) {
            arr[i] = sc.nextInt();
        }

        int index = 1;
        Stack<Integer> stack = new Stack<>();

        for (int i = 0; i < arr.length; i++) {
            if (arr[i] == index) {
                index++;
                continue;
            }

            if (!stack.isEmpty() && stack.peek() == index) {
                stack.pop();
                index++;
                i--;
            } else {
                stack.push(arr[i]);
            }
        }

        boolean isChecked = false;
        while (!stack.isEmpty()) {
            if (stack.peek() != index) {
                isChecked = true;
                break;
            }

            stack.pop();
            index++;
        }

        System.out.println(isChecked ? "Sad" : "Nice");
    }
}

File: 백준/Silver/12789. 도키도키 간식드리미/도키도키 간식드리미.py
import sys

# sys.stdin = open("input.txt", "r")

n = int(sys.stdin.readline().strip())
arr = list(map(int, sys.stdin.readline().split()))

stack = []
index = 1
is_flag = True

for i in arr:
    stack.append(i)

    while stack and stack[-1] == index:
        stack.pop()
        index += 1
    
    if len(stack) > 1 and stack[-1] > stack[-2]:
        is_flag = False
        break

if stack:
    is_flag = False

print('Nice' if is_flag else 'Sad')